### PR TITLE
tbb set_num_threads

### PR DESCRIPTION
**Pull request by cpuhrsch**
This PR hooks up TBB with set_num_threads and also cleans up the code within ReduceOps.cpp a bit.

You can see that the restriction works using the following script

```
import torch
import argparse

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description='Process some integers.')
    parser.add_argument('threads', type=int)
    args = parser.parse_args()

    tv = torch.randn(1000 * 1000 * 10)
    if args.threads > 0:
        torch.set_num_threads(args.threads)
    for _ in range(10000):
        tv.sum()
```

```
(base) [16:44:39: cpuhrsch@devfair0129 benchmarks]$ perf stat python sum_stress.py 5

 Performance counter stats for 'python sum_stress.py 5':

      17327.154266      task-clock (msec)         #    3.721 CPUs utilized          
               962      context-switches          #    0.056 K/sec                  
                90      cpu-migrations            #    0.005 K/sec                  
            22,517      page-faults               #    0.001 M/sec                  
    38,071,989,548      cycles                    #    2.197 GHz                    
   <not supported>      stalled-cycles-frontend  
   <not supported>      stalled-cycles-backend   
    31,333,390,416      instructions              #    0.82  insns per cycle        
     3,732,742,496      branches                  #  215.427 M/sec                  
         8,014,202      branch-misses             #    0.21% of all branches        

       4.656548693 seconds time elapsed

(base) [16:44:44: cpuhrsch@devfair0129 benchmarks]$ perf stat python sum_stress.py 2

 Performance counter stats for 'python sum_stress.py 2':

      16708.310926      task-clock (msec)         #    1.780 CPUs utilized          
             1,893      context-switches          #    0.113 K/sec                  
                85      cpu-migrations            #    0.005 K/sec                  
            21,486      page-faults               #    0.001 M/sec                  
    36,683,508,608      cycles                    #    2.196 GHz                    
   <not supported>      stalled-cycles-frontend  
   <not supported>      stalled-cycles-backend   
    30,414,604,047      instructions              #    0.83  insns per cycle        
     3,536,252,326      branches                  #  211.646 M/sec                  
         5,265,418      branch-misses             #    0.15% of all branches        

       9.386758238 seconds time elapsed

(base) [16:51:54: cpuhrsch@devfair0129 benchmarks]$ perf stat python sum_stress.py 0

 Performance counter stats for 'python sum_stress.py 0':

      47343.842407      task-clock (msec)         #   19.877 CPUs utilized          
           135,891      context-switches          #    0.003 M/sec                  
               623      cpu-migrations            #    0.013 K/sec                  
            21,791      page-faults               #    0.460 K/sec                  
   104,047,864,558      cycles                    #    2.198 GHz                    
   <not supported>      stalled-cycles-frontend  
   <not supported>      stalled-cycles-backend   
    46,576,686,318      instructions              #    0.45  insns per cycle        
     7,054,117,917      branches                  #  148.998 M/sec                  
        33,104,250      branch-misses             #    0.47% of all branches        

       2.381825636 seconds time elapsed

(base) [16:52:18: cpuhrsch@devfair0129 benchmarks]$ perf stat python sum_stress.py 1

 Performance counter stats for 'python sum_stress.py 1':

      16819.557761      task-clock (msec)         #    0.991 CPUs utilized          
               905      context-switches          #    0.054 K/sec                  
                79      cpu-migrations            #    0.005 K/sec                  
            23,798      page-faults               #    0.001 M/sec                  
    36,957,423,434      cycles                    #    2.197 GHz                    
   <not supported>      stalled-cycles-frontend  
   <not supported>      stalled-cycles-backend   
    30,260,079,596      instructions              #    0.82  insns per cycle        
     3,503,067,589      branches                  #  208.273 M/sec                  
         5,049,148      branch-misses             #    0.14% of all branches        

      16.969856749 seconds time elapsed
```

### CODE REVIEW

File: aten/src/ATen/CPUGeneral.cpp [hunk 1]
```diff
@@ -0,0 +1,14 @@
+#include <ATen/CPUGeneral.h>
+
+namespace at {
+size_t num_threads = 0;
+std::shared_ptr<tbb::task_scheduler_init> tbbinit;
+void set_num_threads(size_t num_threads_) {
+  if (num_threads == 0) {
```
Comment (fmassa) [inline aten/src/ATen/CPUGeneral.cpp]: So we can only initialize the number of threads once?

Comment (cpuhrsch) [reply aten/src/ATen/CPUGeneral.cpp]: Hm, let me look into that.

File: aten/src/ATen/CPUGeneral.cpp [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+    num_threads = num_threads_;
+    tbbinit = std::make_shared<tbb::task_scheduler_init>(num_threads);
+  }
+}
+
+size_t get_num_threads() { return num_threads; }
```
Comment (apaszke) [inline aten/src/ATen/CPUGeneral.cpp]: This should probably be part of the ATen context

Comment (cpuhrsch) [reply aten/src/ATen/CPUGeneral.cpp]: Do you mean part of the at namespace?

**Comment by cpuhrsch** [comment]
Added support for changing number of threads at runtime. Consider the following script

```
import torch
import time
import gc
import argparse

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description='Process some integers.')
    parser.add_argument('threads', type=int)
    parser.add_argument('--decrease', action='store_true')
    args = parser.parse_args()

    tv = torch.randn(1000 * 1000 * 10)
    num_thread = args.threads
    torch.set_num_threads(num_thread)
    tstart = time.time()
    for i in range(100000):
        tv.sum()
        if args.decrease and i % 10000 == 0:
            gc.collect()
            elapsed = time.time() - tstart
            tstart = time.time()
            num_thread = num_thread / 2
            print("decreasing to: " + str(num_thread) + " prev elapsed: " + str(elapsed))
            torch.set_num_threads(num_thread)
```

And output

```
decreasing to: 16 prev elapsed: 0.0332319736481
decreasing to: 8 prev elapsed: 1.24536299706
decreasing to: 4 prev elapsed: 2.19691586494
decreasing to: 2 prev elapsed: 4.26764798164
decreasing to: 1 prev elapsed: 7.82375693321
decreasing to: 0 prev elapsed: 15.4659228325
decreasing to: 0 prev elapsed: 0.84959602356
decreasing to: 0 prev elapsed: 0.842261075974
decreasing to: 0 prev elapsed: 0.804984092712
decreasing to: 0 prev elapsed: 0.823231220245

 Performance counter stats for 'python sum_stress.py 32 --decrease':

     354283.766260      task-clock (msec)         #    9.613 CPUs utilized          
           581,664      context-switches          #    0.002 M/sec                  
             4,625      cpu-migrations            #    0.013 K/sec                  
            49,221      page-faults               #    0.139 K/sec                  
   779,103,968,909      cycles                    #    2.199 GHz                    
   <not supported>      stalled-cycles-frontend  
   <not supported>      stalled-cycles-backend   
   395,023,334,546      instructions              #    0.51  insns per cycle        
    56,253,880,001      branches                  #  158.782 M/sec                  
       165,331,982      branch-misses             #    0.29% of all branches        

      36.854981344 seconds time elapsed
```

**Comment by soumith** [comment]
can you also make tbb / set_num_threads respect `OMP_NUM_THREADS` env variable. It's quite common to use it.

File: aten/src/ATen/CPUGeneral.cpp (hunk 2 above)
Comment (apaszke) [reply aten/src/ATen/CPUGeneral.cpp]: No I meant the `at::Context` object. That's what we use to hold all other parts of the CPU/CUDA state

Comment (apaszke) [reply aten/src/ATen/CPUGeneral.cpp]: You probably won't need `shared_ptr` once you move it as well

File: aten/src/ATen/native/ReduceOps.cpp [hunk 3]
```diff
@@ -15,122 +15,103 @@
...
-Tensor sum(const Tensor &self, int64_t dim_, bool keepdim) {
+static Tensor _dimreduce_setup(const Tensor &self, int64_t dim_,
+                               int64_t ident) {
   int64_t dim = maybe_wrap_dim(dim_, self.dim());
   Tensor result = self.type().tensor();
   if (self.numel() == 1 && self.ndimension() == 0) {
     result.resize_({});
     result.fill_(self);
-    return result;
+    return std::move(result);
```
Comment (apaszke) [inline aten/src/ATen/native/ReduceOps.cpp]: Never `return std::move(...)`. This prevents return value optimization and ends up in slower code (clang even warns about it)

File: aten/src/ATen/native/ReduceOps.cpp [hunk 4]
```diff
@@ -15,122 +15,103 @@
...
-    result.resize_({});
-    result.fill_(self);
-    return result;
-  }
-  // Return identity
-  if (self.numel() == 0 && self.ndimension() == 1) {
-    result.resize_({0});
-    result.fill_(1);
+Tensor _prod_cpu(const Tensor &self) {
+  return _reduce_cpu(prodImpl, self);
```
Comment (apaszke) [inline aten/src/ATen/native/ReduceOps.cpp]: This will fall back to `self._sumall` if the `if` fails

File: aten/src/ATen/native/ReduceOps.cpp [hunk 5]
```diff
@@ -15,122 +15,103 @@
...
+}
+
+Tensor &_reduce_out_cpu(reduce_type *f, Tensor &result, const Tensor &self, int64_t dim,
+                     bool keepdim) {
+  if (self.is_contiguous() && result.is_contiguous()) {
+    f(result, self, dim, false);
+    if (!keepdim)
+      result.squeeze_(dim);
     return result;
   }
```
Comment (apaszke) [inline aten/src/ATen/native/ReduceOps.cpp]: This looks like it's missing a `return` if the `if` condition fails

File: aten/src/ATen/native/cpu/Parallel.h [hunk 6]
```diff
@@ -7,18 +8,18 @@
 namespace at {
 namespace native {
 
-constexpr size_t _THRESHOLD = 32768;
+constexpr size_t GRAIN_SIZE = 32768;
```
Comment (apaszke) [inline aten/src/ATen/native/cpu/Parallel.h]: Can you please add a comment for this constant?

Comment (cpuhrsch) [reply aten/src/ATen/native/cpu/Parallel.h]: Sure. I'll revise the documentation for the entire setup in another PR as well. In particular the README file.

File: aten/src/ATen/CPUGeneral.cpp (hunk 2 above)
Comment (colesbury) [reply aten/src/ATen/CPUGeneral.cpp]: I don't really see how moving this to at::Context improves the API.

`at::set_num_threads()` is both simpler to use and matches the Python API more closely than `at::globalContext().set_num_threads()`.

Putting this in the context seems to imply a false sense of isolation: Multiple task_scheduler_inits will still interfere in separate contexts.

Comment (apaszke) [reply aten/src/ATen/CPUGeneral.cpp]: I'm just saying that this is where we usually keep all our ATen state, and don't have any other global settings at the moment. If they would somehow interfere then sure, making it global is ok, but I can't see why they would.

**Comment by cpuhrsch** [comment]
We came upon one issues related to std::thread and tbb. When running e.g. sum within a std::thread, the number of threads chosen will not be respected. So, even if I launch the problem with OMP_NUM_THREADS=1, the std::thread will use the default (all available cores). This causes issues with autograd (engine.cpp). The proposed solution is to only set the number of threads and then have this global variable be read by the respective threads. Then you create a static tbb init object within each parallel function that will initialize itself with the current number of threads. This object is updated only if the number of threads has been changed. There does not appear to be any noticeable performance penalty for this and it also appears to resolve the issue.

File: aten/src/ATen/native/cpu/Parallel.h [hunk 7]
```diff
@@ -7,18 +8,33 @@
...
                   size_t start, size_t end, T init_) {
+
+  static tbb::task_scheduler_init tbbinit;
+  static int num_threads_ = 0;
+  int num_threads = at::get_num_threads();
+  if (num_threads_ != num_threads) {
+    if (tbbinit.is_active()) tbbinit.terminate();
+    tbbinit.initialize(num_threads);
+    num_threads_ = num_threads;
+  }
```
Comment (apaszke) [inline aten/src/ATen/native/cpu/Parallel.h]: This is not thread safe.

Comment (apaszke) [reply aten/src/ATen/native/cpu/Parallel.h]: Does `task_scheduler_init` actually keep the threads alive, or is it just a small struct holding the desired config?

Comment (cpuhrsch) [reply aten/src/ATen/native/cpu/Parallel.h]: It initializes the thread scheduler.

Comment (apaszke) [reply aten/src/ATen/native/cpu/Parallel.h]: Meaning that its creation spawns the threads? Is the scheduler global?

Comment (cpuhrsch) [reply aten/src/ATen/native/cpu/Parallel.h]: At least the creation of the static local variables should be [thread safe](http://en.cppreference.com/w/cpp/language/storage_duration) since C++11. I'll guard the case when num_threads_ is not equal to num_threads.

Comment (cpuhrsch) [reply aten/src/ATen/native/cpu/Parallel.h]: The most important part is that tbbinit is in scope of the parallel constructs. That's what actually sets the number of threads used for the overall [market](https://software.intel.com/en-us/blogs/2011/04/09/tbb-initialization-termination-and-resource-management-details-juicy-and-gory).

File: aten/src/ATen/native/cpu/Parallel.h [hunk 8]
```diff
@@ -7,18 +8,33 @@
...
+// deemed inefficient to parallelise over arrays shorter than 32768. Further,
+// no parallel algorithm (such as parallel_reduce) should split work into
+// smaller than GRAIN_SIZE chunks.
+constexpr size_t GRAIN_SIZE = 32768;
 
 template <class T, template <class> class PRED>
 T parallel_reduce(T (*f)(const T *, size_t, size_t, T), const T *data,
                   size_t start, size_t end, T init_) {
+
+  static tbb::task_scheduler_init tbbinit;
```
Comment (colesbury) [inline aten/src/ATen/native/cpu/Parallel.h]: This doesn't look right. Each instantiation of `parallel_reduce` will have its own copy of `tbb::task_scheduler_init tbbinit`.

I still don't fully understand how `tbb::task_scheduler_init` works, but I don't think these will stack properly.

Comment (apaszke) [reply aten/src/ATen/native/cpu/Parallel.h]: Also, it looks like `task_scheduler_init` is mostly meant to initialize some thread-local variables, so making it `static` might not be a good idea.

Comment (cpuhrsch) [reply aten/src/ATen/native/cpu/Parallel.h]: What could be a better approach?

Comment (cpuhrsch) [reply aten/src/ATen/native/cpu/Parallel.h]: Using thread_local static variables seem to do the trick. The thread_local keyword makes sure that there will be a new variable per thread and also removes any need to guard for data races. The static variable makes sure it's linked internally. 

On task_scheduler_init, it needs to be called before the first usage of any tbb function. It initializes the scheduler with the passed parameters (number of threads). Any tbb construct (such as the parallel algorithm) will spawn a default thread pool if this hasn't been called before. The default will consume all available cores. See [the documentation](https://www.threadingbuildingblocks.org/docs/help/reference/task_scheduler/task_scheduler_init_cls.html):

"To override the automatic defaults for task scheduling, a task_scheduler_init must become active before the first use of task scheduling services.
"
There are two potential issues:
We must direct any usage of tbb to Parallel.h to make sure tbb is initialized first.
While the current scheduler might be limited in the number of cores used, the overall market is not. That means, if there are two threads, they each will spawn the number of threads as specified by torch.set_num_thread. As far as I know this behavior is consistent with OpenMP.

One advantage, as far as I can tell, is that tbb will use a market to limited the number of overall worker threads. So that even if each cpp thread requests 60 threads for each tbb scheduler, the overall number of workers will be limited to the maximum number of cores available and the associated number of worker threads is divided by the number of active schedulers. This should prevent oversubscription. 

See https://software.intel.com/en-us/blogs/2011/04/09/tbb-initialization-termination-and-resource-management-details-juicy-and-gory

**Review by colesbury — APPROVED** [review]
